Add ClojureScript support

jeisses opened the pull request:
Replace functions that are not supported in Quil clojurescript.
Also update the clojure, clojurescript and quil dependencies

Commit:
Add ClojureScript support

Replace functions that are not supported in Quil clojurescript.
Also update the clojure, clojurescript and quil dependencies

echeran commented:
Cool!  I definitely can use help with ClojureScript-ifying things, so this is greatly appreciated!  (Help in this department will turn CLJS support from a wish list to a nearer-term work list item.)  I'll take a look at the PR in further detail later today.

For now, I'll just leave some initial questions/comments.  Also, is there an easy way to give the ClojureScript version of the code a spin?  (Or is that an item for the todo list?)

echeran commented on project.clj:
```diff
@@ -11,6 +11,7 @@
-  :dependencies [[org.clojure/clojure "1.6.0"]
-                 [quil "2.2.5"]]
+  :dependencies [[org.clojure/clojure "1.7.0"]
+                 [org.clojure/clojurescript "1.7.170"]
```
Once we start using CLJS, should we either include lein-cljsbuild or a build.clj for building the CLJS?

(Thanks for bumping the versions on the deps.  And getting Clojure 1.7.0 to use reader conditionals.)

jeisses replied:
I don't think lein-cljsbuild dep is required in the library. I currently require lein-cljsbuild only in my project (which has clojure-turtle as dependency), and the clojure-turtle dependency is build automatically. 

Edit: maybe include lein-cljsbuild in the dev profile for easy testing

echeran commented on src/clojure_turtle/core.cljc:
```diff
@@ -231,28 +230,39 @@
+  (q/translate (/ (q/width) 2) (/ (q/height) 2))
+  (reset-rendering)
+  (q/push-matrix)
+  (q/scale 1.0 -1.0)
```
the apply-matrix call in the current code will adjust the coordinates so that we use a math-style grid (positive x axis goes right, positive y axis goes up).  is that what the scale call achieves?

out of curiosity, this diff hunk of new code seems to be functionally equivalent to the old -- is the new code there to support CLJS in a way that the old doesn't?

jeisses replied:
The **apply-matrix** and **with-translation** function/macro both aren't available for clojurescript. So I replaced them with the equivalent(?) q/translate and q/scale.

The result should be identical, so this diff is for cljs compatibility only.

echeran replied:
Yup, looks good to me -- the CLJS-friendly graphics canvas transformation code here is still functionally equivalent to the existing code.

jeisses commented:
Hi, nice to hear you're interested in clojurescript support! 

The changes are currently functional, I only had to replace functions that ProcessingJS doesn't support. And added the cljs defsketch macro. 
This works really nice with Figwheel, you can live control the turtle in your browser from the REPL. I'll commit some instructions soon on how to get it running so you can test it.

Edit: i'll also add some cljs build config to the dev profile of this project

Commit:
Create a CLJS canvas and bind to Quil applet

A HTML5 canvas object gets created in the new-window macro and is set as
the main Quil *applet*.

Commit:
Add a CLJS demo with Figwheel support

jeisses commented:
So there were a few hurdles after all :) 
In order to keep the repl plug-and-play nature of the new-window macro a few additions were needed for clojurescript:
- A canvas object is now inserted into the DOM. For an unknown reason didn't play nice  with reader conditionals so I ended up using the [if-cljs](http://goo.gl/Bhdyna) macro
- Quil cljs requires all `q/*` calls to be within the applet context, including stuff like `q/radians`. This means the turtle commands won't work from the REPL without wrapping them with `quil.core/with-sketch`, which is troublesome. I found a [solution](https://github.com/jaredly/tutorial-cljs/blob/master/src/tutorial_cljs/quil_sketch.cljs#L184) by binding the global quil `*applet*` in the sketch setup.
- The cljs `defsketch` command must be executed before DOMContentLoaded. This is when ProcessingJS does its init. So atm the defsketch can't be executed from REPL but must be run in a loaded JS file. There are workarounds for this which could be explored.

My last commit contains a demo that can be run with `lein figwheel`; then browse to http://localhost:3449 and use the figwheel repl. What do you think?

echeran commented:
Excellent work!  You did a good job in following all of those fixes and workarounds to get thing working.  It's great to see the canvas in a webpage, now.

The issue of not being able to call `defsketch` from the REPL is not so much of an issue, but it could be useful to think about in the future.  The reason being that truly independent sessions of canvas + turtle state hasn't yet been worked out to begin with... in the Java Clojure version, new windows only reflect the state of the turtle atom in clojure-turtle.core.  Plumbing exists to not rely on such global variables -- you can pass any other turtle atom to the main turtle fns (`forward`, `right`, ...), but that generalization work didn't carry through to sketch rendering, which only uses the single global state turtle atom to paint the canvas.  The tradeoff I was wrestling with is how it would affect the difficulty for users who are new to programming in having to always pass in a turtle atom for each fn (ex: `(forward my-turt 30) (right my-turt 90) ...`), and since Logo doesn't do it, I decided to not pursue promoting that idea.  There's also a limitation in the Clojure/Script implementations in that you only have 1 REPL / Figwheel session.  So with only prompt open, it's simplest to only allow the user to manipulate one turtle's state, anyways.

If there's a utility to `new-window` from the Clojure REPL, it's that you can obtain the Quil canvas size of your choosing (ex: the turtle "goes off the screen", but what the turtle drew still exists and can be seen with a bigger window).  So if you explore a way to call `defsketch` after ProcessingJS has loaded, if that's possible, then maybe an HTML form next to the canvas that allows the user to enter the canvas size and have that take effect.  (On a side note, what I thought was interesting was what happens when you load the Figwheel URL into multiple tabs at different times.  Each tab's canvas's turtle responds to Figwheel's REPL session, but each turtle may have different states depending on whether they received earlier commands or were started too late for that.)

One thing that I've thought of as another way to handle multiple turtle sessions is do so serially, by allowing the user to save and load the state.  I can think of how to do that in plain Clojure, but I don't know if ClojureScript supports `pr`/`pr-str` or `clojure.edn/read`.  And although it might take a bit of effort and time, another idea that CLJS could use is to embed a REPL in the webpage that connects to the canvas of the same page, sort of like what Replumb (demo'ed on http://clojurescript.io).
